File: proto_checker.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
__name__ = "ProtoChecker"
__author__ = "VegaS"
__date__ = "2019-11-18"
__version__ = "0.0.3"

import sys
import os
import string
import csv

# Set the base color of command prompt
os.system("color 00")

# Set working directory
os.chdir(os.path.dirname(os.path.realpath(__file__)))

# Auto detection of vnum range
VNUM_RANGE_ITEM_VNUM_START = 110000 # Start item vnum
VNUM_RANGE_ITEM_VNUM_END = 165400 # End item vnum
VNUM_RANGE = 99 # itemVnumStart~itemVnumEnd+99
VNUM_RANGE_DELIMITER = '~' # Separator

#################################################
## Builtin translations
#################################################
COLOR_GREEN = "\033[92m{}\033[00m"
COLOR_RED = "\033[91m{}\033[00m"
COLOR_GRAY = "\033[37m{}\033[00m"

TRANSLATE_DICT = {
	"FILE_EMPTY": COLOR_RED.format("File {} is empty."),
	"FILE_NAME": COLOR_GRAY.format("Reading file: {}"),
	"FILE_OK": COLOR_GREEN.format("\tOK"),
	"FILE_TOTAL_LINES": COLOR_RED.format("\tTotal lines: {}"),
	"FILE_DUPLICATE_LINE": COLOR_RED.format("\tDuplicated itemVnum: {} at line: {}"),
	"FILE_COMPARING_LINE": COLOR_RED.format("\tMissing itemVnum: {}"),
}

for localeName, localeValue in TRANSLATE_DICT.items():
	locals().update({localeName: localeValue})

#################################################
## ProtoChecker
#################################################
class ProtoChecker:
	BASE_PATH = "resource\\"
	LOG_FILE_NAME = 'syslog.txt'
	
	ITEM_PROTO_FILE_NAME = BASE_PATH + 'item_proto.txt'
	ITEM_NAMES_FILE_NAME = BASE_PATH + 'item_names.txt'
	MOB_PROTO_FILE_NAME = BASE_PATH + 'mob_proto.txt'
	MOB_NAMES_FILE_NAME = BASE_PATH + 'mob_names.txt'

	def __init__(self):
		self.outputFileList = []

		self.itemProtoDict = self.load_file(self.ITEM_PROTO_FILE_NAME)
		self.itemNamesDict = self.load_file(self.ITEM_NAMES_FILE_NAME)

		self.mobProtoDict = self.load_file(self.MOB_PROTO_FILE_NAME)
		self.mobNamesDict = self.load_file(self.MOB_NAMES_FILE_NAME)

	def load_file(self, fileName):
		fileExists = os.path.exists(fileName)
		if fileExists:
			if os.path.getsize(fileName) == 0:
				self.write_log(FILE_EMPTY.format(fileName))
				self.create_syslog()
				sys.exit(0)

		fileData = []
		vnumRangeSet = set()
		if fileExists:
			with open(fileName, 'r') as csv_file:
				csv_reader = csv.reader(csv_file, delimiter='\t')
				next(csv_reader, None)

				for row in csv_reader:
					if row:
						column = row[0]
						if VNUM_RANGE_DELIMITER in column:
							vnumRangeSet.add(tuple(map(int, column.split(VNUM_RANGE_DELIMITER))))

						fileData.append(column)

		fileData.sort()
		return {'name': fileName, 'data': fileData, 'range': vnumRangeSet, 'exists': fileExists}

	def find_duplicate(self, file):
		if not file["exists"]:
			return

		self.write_log(FILE_NAME.format(file["name"]))

		duplicateSet = set([itemVnum for itemVnum in file["data"] if file["data"].count(itemVnum) > 1])
		for itemVnum in duplicateSet:
			self.write_log(FILE_DUPLICATE_LINE.format(itemVnum, len(file["data"]) - file["data"][-1::-1].index(itemVnum) + 1))

		self.write_log(FILE_TOTAL_LINES.format(duplicateSet.__len__()) if duplicateSet else FILE_OK)

	# noinspection PyDefaultArgument
	def compare(self, firstFile, secondFile):
		differenceList = []

		def check_diff(value, data):
			if value not in data:
				differenceList.append(value)

		if not firstFile['exists'] or not secondFile['exists']:
			return

		self.write_log(FILE_NAME.format(secondFile['name']))

		cmpItemProto = secondFile['name'] == self.ITEM_PROTO_FILE_NAME
		cmpItemNames = secondFile['name'] == self.ITEM_NAMES_FILE_NAME
		if cmpItemProto or cmpItemNames:
			for itemVnum in firstFile['data']:
				baseItemVnum = itemVnum

				# Comparing item_names with item_proto
				if cmpItemProto:
					if not itemVnum.isdigit():
						continue

					intItemVnum = int(itemVnum)
					if intItemVnum in range(VNUM_RANGE_ITEM_VNUM_START, VNUM_RANGE_ITEM_VNUM_END + 1):
						itemVnumRange = '{start}{separator}{end}'.format(start=intItemVnum, separator=VNUM_RANGE_DELIMITER, end=intItemVnum + VNUM_RANGE)
						baseItemVnum = itemVnumRange

					check_diff(baseItemVnum, secondFile['data'])

				# Comparing item_proto with item_names
				elif cmpItemNames:
					if VNUM_RANGE_DELIMITER in itemVnum:
						itemVnumStart, itemVnumEnd = itemVnum.split(VNUM_RANGE_DELIMITER)
						baseItemVnum = itemVnumStart
						
					check_diff(baseItemVnum, secondFile['data'])
		else:
			# Comparing mob_proto with mob_names and vice-versa
			differenceList = list(set(firstFile['data']).difference(secondFile['data']))

		if not differenceList:
			self.write_log(FILE_OK)
			return

		for itemVnum in differenceList:
			self.write_log(FILE_COMPARING_LINE.format(itemVnum))

		self.write_log(FILE_TOTAL_LINES.format(differenceList.__len__()))

	def write_log(self, line, lineDelimiter='\n', formatDelimiter='{}'):
		sys.stdout.write(line + lineDelimiter)

		for color in (COLOR_GREEN, COLOR_RED, COLOR_GRAY):
			for item in color.split(formatDelimiter):
				line = line.replace(item, string.whitespace[0])
		self.outputFileList.append(line)

	def create_syslog(self):
		with open(self.LOG_FILE_NAME, 'w') as out:
			out.write('\n'.join(self.outputFileList))

	def run(self):
		self.write_log("###### START_CHECKING_FOR_DUPLICATE ######")
		for file in (self.itemProtoDict, self.itemNamesDict, self.mobProtoDict, self.mobNamesDict):
			self.find_duplicate(file)
		self.write_log("###### END_CHECKING_FOR_DUPLICATE ######\n")

		self.write_log("###### START_COMPARING ######")
		for files in ((self.itemProtoDict, self.itemNamesDict), (self.mobProtoDict, self.mobNamesDict)):
			for firstFile, secondFile in zip(files, reversed(files)):
				self.compare(firstFile, secondFile)

		self.write_log("###### END_COMPARING ######")
		self.create_syslog()

protoChecker = ProtoChecker()
protoChecker.run()

File: c-source/ProtoChecker/ProtoChecker.cpp
/***********************************
__name__    = "ProtoChecker"
__author__  = "VegaS"
__date__    = "2019-12-05"
__version__ = "0.0.1"
***********************************/
#include "pch.h"
#include "ProtoChecker.h"
#include "CsvReader.h"

/**
 * \brief
 * Character delimiters.
 */
enum EDelimiter : char
{
	DELIMITER_RANGE = '~',
	DELIMITER_TAB	= '\t',
};

/**
 * \brief
 * Vnum range config.
 */
enum EVnumRangeConfig : uint32_t
{
	VNUM_RANGE_START    = 110000,
	VNUM_RANGE_END      = 165400,
	VNUM_RANGE          = 99,
};

/**
 * \brief
 * Path configuration.
 */
static const auto ROOT_PATH             = "resource\\";
static const auto LOG_FILE_NAME         = "syslog.txt";

static const auto ITEM_PROTO_FILE_NAME  = "item_proto.txt";
static const auto ITEM_NAMES_FILE_NAME  = "item_names.txt";
static const auto MOB_PROTO_FILE_NAME   = "mob_proto.txt";
static const auto MOB_NAMES_FILE_NAME   = "mob_names.txt";

/**
 * \brief
 * Builtin translation.
 */
using TLocaleStringMap = std::map<std::string, std::string>;
TLocaleStringMap TRANSLATE_MAP =
{
	{"FILE_EMPTY",          format(COLOR_GREEN, "File %s is empty.")},
	{"FILE_OK",             format(COLOR_GREEN, "\tOK")},
	{"FILE_NAME",           format(COLOR_GRAY,  "Reading file: %s")},
	{"FILE_TOTAL_LINES",    format(COLOR_RED,   "\tTotal lines: %d")},
	{"FILE_DUPLICATE_LINE", format(COLOR_RED,   "\tDuplicated itemVnum: %s at line: %d")},
	{"FILE_COMPARING_LINE", format(COLOR_RED,   "\tMissing itemVnum: %s")},
};

/**
 * \brief
 * Called the constructor/destructor method.
 */
CProtoChecker::CProtoChecker() :
	m_ItemProtoFile(load_file(ITEM_PROTO_FILE_NAME)),
	m_ItemNamesFile(load_file(ITEM_NAMES_FILE_NAME)),
	m_MobProtoFile(load_file(MOB_PROTO_FILE_NAME)),
	m_MobNamesFile(load_file(MOB_NAMES_FILE_NAME))
{
}

CProtoChecker::~CProtoChecker()
{
}

/**
 * \brief
 * Load a specific file and save the data into a tuple.
 * \param: stFileName: string
 * \return: tuple
 */
auto CProtoChecker::load_file(const std::string& stFileName) const -> TFileTuple
{
	cCsvTable csv_reader;
	std::vector<std::string> vecFileData;
	std::vector<std::tuple<uint32_t, uint32_t>> vecRangeVnum;

	std::string stFilePath(ROOT_PATH);
	stFilePath.append(stFileName);

	const auto bFileExists = csv_reader.Load(stFilePath.c_str(), DELIMITER_TAB);
	if (bFileExists)
	{
		csv_reader.Next();
		while (csv_reader.Next())
		{
			const auto & stItemVnum = csv_reader.AsStringByIndex(0);
			if (stItemVnum.find(DELIMITER_RANGE) != std::string::npos)
			{
				const auto & vecVnumRange = split_range(stItemVnum, std::string(1, DELIMITER_RANGE));
				vecRangeVnum.emplace_back(strtoul(vecVnumRange[0]), strtoul(vecVnumRange[1]));
			}

			vecFileData.emplace_back(stItemVnum);
		}
	}

	return std::make_tuple(stFileName, vecFileData, bFileExists, vecRangeVnum);
}

/**
 * \brief
 * Read a specific file and find the duplicates lines.
 * \param: file: tuple
 */
auto CProtoChecker::find_duplicate(const TFileTuple& file) -> void
{
	if (!std::get<EXISTS>(file))
		return;

	const auto & stFileName = std::get<NAME>(file);
	const auto & vecFileData = std::get<DATA>(file);

	write_log(TRANSLATE_MAP["FILE_NAME"], stFileName.c_str());

	std::map<std::string, uint32_t> mapFileDuplicate;
	uint32_t lineIndex = 1;
	for (const auto & stItemVnum : vecFileData)
	{
		++lineIndex;
		const auto iCount = std::count(vecFileData.begin(), vecFileData.end(), stItemVnum);
		if (iCount > 1)
			mapFileDuplicate[stItemVnum] = lineIndex;
	}

	for (const auto & it : mapFileDuplicate)
		write_log(TRANSLATE_MAP["FILE_DUPLICATE_LINE"], it.first.c_str(), it.second);

	write_log(mapFileDuplicate.empty() ? TRANSLATE_MAP["FILE_OK"] : TRANSLATE_MAP["FILE_TOTAL_LINES"], mapFileDuplicate.size());
}

/**
 * \brief
 * Compare two files and find the differences between them.
 * \param: fileExamine: tuple
 * \param: fileSearch: tuple
 */
auto CProtoChecker::compare(const TFileTuple& fileExamine, const TFileTuple& fileSearch) -> void
{
	if (!std::get<EXISTS>(fileExamine) || !std::get<EXISTS>(fileSearch))
		return;
	 
	auto vecFileExamine = std::get<DATA>(fileExamine);
	auto vecFileSearch = std::get<DATA>(fileSearch);
	decltype(vecFileSearch) vecFileDifference;

	std::sort(vecFileExamine.begin(), vecFileExamine.end());
	std::sort(vecFileSearch.begin(), vecFileSearch.end());

	const auto bCmpItemProto = std::get<NAME>(fileSearch) == ITEM_PROTO_FILE_NAME;
	const auto bCmpItemNames = std::get<NAME>(fileSearch) == ITEM_NAMES_FILE_NAME;

	if (bCmpItemProto || bCmpItemNames)
	{
		for (const auto & stItemVnum : vecFileExamine)
		{
			auto stBaseItemVnum = stItemVnum;
			if (bCmpItemProto)
			{
				const auto itemVnum = strtoul(stItemVnum);
				if (itemVnum >= VNUM_RANGE_START && itemVnum <= VNUM_RANGE_END)
				{
					char cVnumRange[64];
					_snprintf_s(cVnumRange, sizeof(cVnumRange), "%u%c%u", itemVnum, DELIMITER_RANGE, itemVnum + VNUM_RANGE);
					stBaseItemVnum = cVnumRange;
				}
			}
			if (bCmpItemNames)
			{
				if (stItemVnum.find(DELIMITER_RANGE) != std::string::npos)
				{
					const auto stItemVnumStart = split_range(stItemVnum, std::string(1, DELIMITER_RANGE)).at(0);
					stBaseItemVnum = stItemVnumStart;
				}
			}
			if (std::find(vecFileSearch.begin(), vecFileSearch.end(), stBaseItemVnum) == vecFileSearch.end())
				vecFileDifference.emplace_back(stBaseItemVnum);
		}
	}
	else
	{
		std::set_difference(vecFileExamine.begin(), vecFileExamine.end(), vecFileSearch.begin(), vecFileSearch.end(), std::inserter(vecFileDifference, vecFileDifference.begin()));
	}

	write_log(TRANSLATE_MAP["FILE_NAME"], std::get<EFile::NAME>(fileSearch).c_str());

	if (vecFileDifference.empty())
	{
		write_log(TRANSLATE_MAP["FILE_OK"]);
		return;
	}

	for (const auto & stItemVnum : vecFileDifference)
		write_log(TRANSLATE_MAP["FILE_COMPARING_LINE"], stItemVnum.c_str());

	write_log(TRANSLATE_MAP["FILE_TOTAL_LINES"], vecFileDifference.size());
}

/**
 * \brief
 * Split a string by a specific regular expression.
 * \param:: stItemVnum: string
 * \param:: stRegex: string
 * \return: vector
 */
auto CProtoChecker::split_range(const std::string& stItemVnum, const std::string& stRegex) -> std::vector<std::string>
{
	try
	{
		const std::regex re(stRegex);
		const std::sregex_token_iterator first{stItemVnum.begin(), stItemVnum.end(), re, -1};
		const std::sregex_token_iterator last;
		return { first, last };
	}
	catch (const std::regex_error & e)
	{
		std::cout << e.what() << e.code() << std::endl;
	}
	return {};
}

/**
 * \brief
 * Strip a specific string from the beginning and at the end of the string.
 * \param: stSrc: string
 * \param: stFrom: string
 */
auto CProtoChecker::strip_color(std::string & stSrc, const std::string& stFrom) -> void
{
	size_t iStartPos = 0;
	while ((iStartPos = stSrc.find(stFrom, iStartPos)) != std::string::npos)
		stSrc.replace(iStartPos, stFrom.length(), "");
}

/**
 * \brief
 * Write a log line to console and specific log file.
 * \param: fmt: string
 * \param: ...: any
 */
auto CProtoChecker::write_log(const std::string fmt, ...) -> void
{
	char cLogLine[512];
	va_list args;
	va_start(args, fmt);
	_vsnprintf_s(cLogLine, sizeof(cLogLine), fmt.c_str(), args);
	va_end(args);

	std::string stLogLine = cLogLine;
	std::cout << stLogLine << std::endl;

	for (const auto & color: {COLOR_GREEN, COLOR_GRAY, COLOR_RED})
	{
		for (const auto & item : split_range(color, "%s"))
			strip_color(stLogLine, item);
	}
	
	m_vecLogFile.emplace_back(stLogLine);
}

/**
 * \brief
 * Create the syslog file.
 */
auto CProtoChecker::create_file_log() -> void
{
	std::ofstream outfile(LOG_FILE_NAME);
	for (const auto & line : m_vecLogFile)
		outfile << line << std::endl;

	outfile.close();
}

/**
 * \brief
 * Run the program.
 */
auto CProtoChecker::run() -> void
{
	write_log("###### START_CHECKING_FOR_DUPLICATE ######");
	for (const auto & file : { m_ItemProtoFile, m_ItemNamesFile, m_MobProtoFile, m_MobNamesFile })
		find_duplicate(file);
	write_log("###### END_CHECKING_FOR_DUPLICATE ######");

	write_log("###### START_COMPARING ######");
	compare(m_ItemProtoFile, m_ItemNamesFile);
	compare(m_ItemNamesFile, m_ItemProtoFile);
	compare(m_MobProtoFile, m_MobNamesFile);
	compare(m_MobNamesFile, m_MobProtoFile);
	write_log("###### END_COMPARING ######");

	create_file_log();
}